tools/docgen/docgen/channels.py

```diff
@@ -41,3 +51,207 @@ def process(self, input_file: str) -> ():
+                with open(img_path, 'wb') as img_file:
+                    img_file.write(content)
+                img_path_rel = os.path.relpath(img_path, relative_to).replace(os.sep, "/")
+                resources_img.append(img_path_rel)
+    
+        img_tags = html_soup.find_all('img')
+        for img_tag in img_tags:
+            img_loc = img_tag['src']
+
+            if self._is_valid_url(img_loc):
+                 # downloaded image
+                response = requests.get(img_loc)
+                if response.status_code == 200:
+                    img_filename = self._sentence_to_snake(img_loc.split("/")[-1])
+                    img_path = os.path.join(output_folder, img_filename)
+                    with open(img_path, 'wb') as img_file:
+                        img_file.write(response.content)
+                    img_path_rel = os.path.relpath(img_path, relative_to).replace(os.sep, "/")
+                    img_tag['src'] = img_path_rel
+                else:
+                    raise ValueError(f"Could not download image from {img_loc}")
+                
+            elif get_image_from_local:
+                # process local images
+                img_filename = self._sentence_to_snake(img_loc.split("/")[-1]).replace("_", "-")
+                file_folder = "/".join(notebook_path.split("/")[:-1]) # path read from manifest file
+                img_input_path = os.path.join(self.input_dir, file_folder, img_loc).replace("/", os.sep)
+                if not os.path.exists(img_input_path):
+                    raise ValueError(f"Could not get image from {img_loc}")
+                img_path = os.path.join(output_folder, img_filename)
+                img_path_rel = os.path.relpath(img_path, relative_to).replace(os.sep, "/")
+                shutil.copy(img_input_path, img_path)
+
+            else:
+                # process image got from notebook resources
+                img_path_rel = resources_img[i]
+                img_tag['src'] = img_path_rel
+                i += 1
+
+            self._replace_img_tag(img_tag, img_path_rel)
+
+        return html_soup
+    
+    def _validate_metadata(self, metadata):
+        required_metadata = [
+            "author",
+            "description",
+            "ms.author",
+            "ms.topic",
+            "title",
+        ]
+        for req in required_metadata:
+            assert req in metadata.keys(), f"{req} is required metadata, please add it to manifest file"
+
+    def _generate_metadata_header(self, metadata):
+        """
+        take a file and the authors name, generate metadata
+        metadata requirements: https://learn.microsoft.com/en-us/contribute/metadata
+        Azure Doc require MS authors and contributors need to make content contributions through the private repository
+        so the content can be staged and validated by the current validation rules. (Jan 4th, 2023)
+        """
+        generated_metadata = "---\n"
```

Comment:
Nit: This func can be simplified. (And mutating a string is a python anti pattern) to remove the string mutation First do the date check and if it’s not there add it to the dict directly. Then convert the dict to a list of lines, then concat the list with a newline. Finally just put the new stuff into the larger template with a format string